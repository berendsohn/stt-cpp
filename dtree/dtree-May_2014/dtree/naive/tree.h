// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
