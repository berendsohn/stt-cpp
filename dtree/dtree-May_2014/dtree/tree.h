// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
