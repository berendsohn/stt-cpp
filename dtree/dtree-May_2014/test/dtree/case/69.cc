#define DTREE_CONFIG EndTree<WithEvert<WithDescValue<dtree::DpAdd<double>, Begin > > >
#include "test/dtree/tree.cc"
