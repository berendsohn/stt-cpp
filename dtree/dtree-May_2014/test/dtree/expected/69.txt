<<<
  1: vL296R-85F0
  2: vL184R302F0 -> 14
  3: vL412R341F0 -> 13
  4: vL176R84F0 -> 5
  5: vL278R293F0 -> 8
  6: vL166R668F0
  7: vL594R-129F0
  8: vL146R128F0
  9: vL361R305F0 -> 4
  10: vL471R44F0
  11: vL486R427F0 -> 6
  12: vL603R150F0 -> 11
  13: vL-39R474F0 -> 5
  14: vL-109R76F0 -> 15
  15: vL693R-39F0 -> 11
>>>
<<<
  @1: vL296R-85F0 f1

    /4: vL-102R-209F0 f1
      \9: vL185R221F0 f0
  @5: vL278R293F0 f1
    \8: vL-132R-165F0 f1

  @7: vL594R-129F0 f0

  @10: vL471R44F0 f0

    /6: vL-320R241F0 f0
  @11: vL486R427F0 f0
      /2: vL293R226F0 f0
    \14: vL-595R-351F0 f1
      \15: vL802R-115F0 f1

 11
  \12: vL117R-277F0 f0

 5
    /3: vL451R-133F0 f0
  \13: vL-317R181F0 f1
>>>
